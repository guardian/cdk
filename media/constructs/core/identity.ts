import { Tags } from "aws-cdk-lib";
import type { IConstruct } from "constructs";

export interface StackStageIdentity {
  stack: string;
  stage: string;
}

export interface AppIdentity {
  app: string;
}

export interface Identity extends StackStageIdentity, AppIdentity {}

export const AppIdentity = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- user defined type guard
  isAppIdentity(props: any): props is AppIdentity {
    return props ? "app" in props : false;
  },

  suffixText(appIdentity: AppIdentity, text: string): string {
    const titleCaseApp = appIdentity.app.charAt(0).toUpperCase() + appIdentity.app.slice(1);
    // CloudFormation Logical Ids must be alphanumeric, so remove any non-alphanumeric characters: https://stackoverflow.com/a/20864946
    const alphanumericTitleCaseApp = titleCaseApp.replace(/[\W_]+/g, "");
    return `${text}${alphanumericTitleCaseApp}`;
  },
  taggedConstruct<T extends IConstruct>(appIdentity: AppIdentity, construct: T): T {
    Tags.of(construct).add("App", appIdentity.app);
    return construct;
  },
};
