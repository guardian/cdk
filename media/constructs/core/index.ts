export * from "./identity";
export * from "./migrating";
export * from "./parameters";
export * from "./stack";
