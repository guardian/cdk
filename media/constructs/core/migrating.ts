export interface GuStaticLogicalId {
  /**
   * The logical ID to use in the synthesised template for this resource.
   */
  logicalId: string;

  /**
   * A short description to help developers understand why this resource's logical ID is being set.
   */
  reason: string;
}
