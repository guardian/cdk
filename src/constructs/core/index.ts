export * from "./identity";
export * from "./migrating";
export * from "./parameters";
export * from "./ssm";
export * from "./stack";
