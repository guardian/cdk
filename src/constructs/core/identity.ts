import type { IConstruct } from "@aws-cdk/core";
import { Tags } from "@aws-cdk/core";

export interface StackStageIdentity {
  stack: string;
  stage: string;
}

export interface AppIdentity {
  app: string;
}

export interface Identity extends StackStageIdentity, AppIdentity {}

export const AppIdentity = {
  suffixText(appIdentity: AppIdentity, text: string): string {
    const titleCaseApp = appIdentity.app.charAt(0).toUpperCase() + appIdentity.app.slice(1);
    return `${text}${titleCaseApp}`;
  },
  taggedConstruct<T extends IConstruct>(appIdentity: AppIdentity, construct: T): T {
    Tags.of(construct).add("App", appIdentity.app);
    return construct;
  },
};
