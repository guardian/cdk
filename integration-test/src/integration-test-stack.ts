import { GuDistributionBucketParameter, GuStack } from "@guardian/cdk/lib/constructs/core";
import type { App } from "@aws-cdk/core";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";

export class IntegrationTestStack extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);
    GuDistributionBucketParameter.getInstance(this);
  }
}
