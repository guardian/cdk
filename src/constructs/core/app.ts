import { Tags } from "aws-cdk-lib";
import { Construct } from "constructs";
import type { Identity } from "./identity";
import type { GuStack } from "./stack";

export class GuApp extends Construct implements Identity {
  public readonly stack: string;
  public readonly stage: string;
  public readonly app: string;

  /**
   * A sanitised version of `app` suitable to use for the logical ID of CloudFormation resources, which must be alphanumeric.
   * Typically, only needed when wanting to create a static (deterministic) logical ID during synthesis.
   *
   * @see https://stackoverflow.com/a/20864946
   */
  public readonly appForLogicalId: string;

  public readonly parent: GuStack;

  // eslint-disable-next-line custom-rules/valid-constructors -- `scope` cannot be `GuApp` as that's recursive
  constructor(scope: GuStack, id: string) {
    super(scope, id);

    this.stack = scope.stack;
    this.stage = scope.stage;
    this.app = id;

    const titleCaseApp = id.charAt(0).toUpperCase() + id.slice(1);
    this.appForLogicalId = titleCaseApp.replace(/[\W_]+/g, "");

    this.parent = scope;

    Tags.of(this).add("App", id);
  }
}
