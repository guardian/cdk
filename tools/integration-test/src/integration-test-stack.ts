import { GuDistributionBucketParameter, GuStack } from "@guardian/cdk/lib/constructs/core";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import type { App } from "aws-cdk-lib";

export class IntegrationTestStack extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);
    GuDistributionBucketParameter.getInstance(this);
  }
}
