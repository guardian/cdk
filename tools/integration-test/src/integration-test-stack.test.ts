import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { App } from "@aws-cdk/core";
import { IntegrationTestStack } from "./integration-test-stack";

describe("The Cdk stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new IntegrationTestStack(app, "cdk", { stack: "integration-test", stage: "PROD" });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
