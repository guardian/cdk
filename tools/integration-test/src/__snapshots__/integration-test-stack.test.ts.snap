// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Cdk stack matches the snapshot 1`] = `
Object {
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
}
`;
